src/server/src/main.rs

```diff
@@ -1,23 +1,215 @@
+{
+    let login_user = body.into_inner();
+    let (username,) = path.into_inner();
+    let map = user_map.lock().unwrap();
+
+    if let Some(user) = map.get(&username) 
+    {
+        if user.password == login_user.password 
+        {
+            // generate new session ID -> spawns a child process for the user session -> stores the session ID and child pid
+            let session_id = Uuid::new_v4().to_string();
+            let session_process = spawn_user_session(username.clone());
+            let mut session_map = session_map.lock().unwrap();
+            session_map.insert(session_id.clone(), session_process);
+            return HttpResponse::Ok().json(session_id); //returns id token
+        }
+    }
+    HttpResponse::BadRequest().body("Invalid username/password combination.")
+}
+
+async fn logout_user_handler(
+    body: web::types::Json<String>,
+    path: web::types::Path<(String,)>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let session_id = body.into_inner();
+    let (username,) = path.into_inner();
+    let mut session_map = session_map.lock().unwrap();
+
+    if session_map.remove(&session_id).is_some() 
+    {
+        HttpResponse::Ok().body("User successfully logged out.")
+    } else 
+    {
+        HttpResponse::BadRequest().body("Invalid sessionId.")
+    }
 }
 
 #[ntex::main]
-async fn main() -> std::io::Result<()> {
-    web::HttpServer::new(|| web::App::new().service(index).service(hello))
-        .bind(("127.0.0.1", 8080))?
-        .run()
-        .await
-}
\ No newline at end of file
+async fn main() -> std::io::Result<()> 
+{
+    Bastion::init();
+    Bastion::start();
+
+    // sharing state
+    let user_map = web::types::Data::new(Arc::new(Mutex::new(HashMap::new())));
+    let session_map = web::types::Data::new(Arc::new(Mutex::new(HashMap::new())));
+
+    //hardcoded test user
+    {
+        let mut users = user_map.lock().unwrap();
+        users.insert(
+            "user1".to_string(),
+            User {
+                username: "user".to_string(),
+                password: "password".to_string(),
+            },
+        );
+    }
+
+    HttpServer::new(move || 
+    {
+        App::new()
+            .app_data(user_map.clone()) 
+            .app_data(session_map.clone()) 
+            .service(
+                web::resource("/users/{username}")
+                    .route(web::post().to(create_user_handler)) // Create user
+                    .route(web::get().to(get_user_handler)) // Get user
+                    .route(web::put().to(update_user_handler)) // Update user
+            )
+            .service(
+                web::resource("/users/{username}/login")
+                    .route(web::post().to(login_user_handler)) // Login user
+            )
+            .service(
+                web::resource("/users/{username}/logout")
+                    .route(web::post().to(logout_user_handler)) // Logout user
+            )
```

Comment:
i would probably update the function names to reflect the rest api
`create_user_handler -> post_user_handler`
`put_user_handler    -> put_user_handler`
`login_user_handler  -> post_user_login_handler`
`logout_user_handler -> post_user_logout_handler`
reason being, when im reading the function definition, i don't really want to go and search for what function of the rest api it's implementing. i kind of want that context to be embedded into the name.